File: src/Solutions/Euler090.hs
module Solutions.Euler090 where
import Utils.List(combinations)
import Data.List

squares = map (^2) [1..9]
rev c = if c `elem` [6, 9] then [6, 9] else [c]

check :: [Int] -> [Int] -> Bool
check dice1 dice2 = all (`elem` l) squares where
    l = concat [[10 * n1 + n2, 10 * n2 + n1] | n1 <- concatMap rev dice1, n2 <- concatMap rev dice2]

count = length [() | 
    dice1 <- combinations 6 [0..9],
    dice2 <- combinations 6 [0..9],
    check dice1 dice2] `div` 2

euler090 :: IO String
euler090 = return $ show count

File: rust/src/solutions/euler090.rs
use itertools::Itertools;

const SQUARES: [usize; 9] = [1, 4, 9, 16, 25, 36, 49, 64, 81];

fn rev(n: &usize) -> Vec<usize> {
    if [6, 9].contains(n) {
        vec![6, 9]
    } else {
        vec![*n]
    }
}

fn check(dice1: &Vec<usize>, dice2: &Vec<usize>) -> bool {
    let digits1: Vec<usize> = dice1.iter().flat_map(rev).collect();
    let digits2: Vec<usize> = dice2.iter().flat_map(rev).collect();
    let numbers: Vec<usize> = digits1
        .into_iter()
        .cartesian_product(digits2.into_iter())
        .flat_map(|(a, b)| vec![10 * a + b, 10 * b + a])
        .collect();
    SQUARES.iter().all(|s| numbers.contains(s))
}

pub fn euler090() -> String {
    ((0..10)
        .combinations(6)
        .cartesian_product((0..10).combinations(6))
        .filter(|(a, b)| check(a, b))
        .count()
        / 2)
    .to_string()
}
